### ISSUE TITLE
Add config file for formatting defaults

### ISSUE THREAD

**Issue by mheiber** [dhall-lang/dhall-lang#1086]
## Problem

That the formatter defaults to non-ascii is a barrier to adoption for my team- the pretty lambdas and arrows are perceived as being cute at the expense of practicality, giving the impression that Dhall chooses cuteness over practicality generally. Which isn't true! Dhall seems practicality-oriented once you look past the cute syntax.

## Proposal

- dhall-config.dhall is a file for configuring dhall
- starts with just one option, nested under `formatting`:
    - { formatting: { ascii: Boolean } } -- defaults to false
- dhall discovers these config files somehow (TBD)

> Thanks for this wonderful language. I could take a stab at a PR if this feature seems useful and doesn't detract from the current minimalism and simplicity too much.

**Comment by Gabriel439** [comment]
@mheiber: Yeah, I would accept a pull request for this.  Another useful option to add would be enabling or disabling the `--explain` mode

**Comment by sjakobi** [comment]
> We could add a setting for --ascii to the vscode extension.

https://github.com/dhall-lang/vscode-language-dhall/issues/6 is related.

**Comment by SiriusStarr** [comment]
Hypothetical questions:

Should this be a dotfile, e.g. `.dhall-config.dhall`, since that is often but certainly not always the convention?

Should this also control e.g. `--compact` for `dhall-to-json`?  At the risk of bikeshedding, it would seem prudent to think a bit about how this file would control things beyond formatting, since there are a lot of tools in the Dhall ecosystem, and being able to control them in a relatively standardized fashion would seem preferable to an adhoc solution.

I suppose this matters less if the file is (presumably) going to be in Dhall itself, since anyone can just look at the type then (which would presumably be part of the dhall-haskell repo, if it's specific to that tool; this doesn't seem like part of the standard to me)?

**Comment by philandstuff** [comment]
Opinion: it should be in `$XDG_CONFIG_HOME` somewhere, such as `~/.config/dhall-haskell.dhall`. We already use the XDG spec to define where the cache goes.

**Comment by SiriusStarr** [comment]
> Opinion: it should be in `$XDG_CONFIG_HOME` somewhere, such as `~/.config/dhall-haskell.dhall`. We already use the XDG spec to define where the cache goes.

I think the idea was to be able to set it more on a per-project basis.  If you want to set it system-wide, you can do that easily enough with e.g. `alias dhall-format="dhall format --ascii"`.  Since the idea was to synchronize settings with a team, you'd presumably want the file to be part of the repo, rather than in `$XDG_CONFIG_HOME` or wherever.

**Comment by german1608** [comment]
> Opinion: it should be in `$XDG_CONFIG_HOME` somewhere, such as `~/.config/dhall-haskell.dhall`. We already use the XDG spec to define where the cache goes.

Agree. `dhall-docs` already uses `$XDG_CONFIG_HOME` for the generated docs as well.

---

> I think the idea was to be able to set it more on a per-project basis. If you want to set it system-wide, you can do that easily enough with e.g. `alias dhall-format="dhall format --ascii"`. Since the idea was to synchronize settings with a team, you'd presumably want the file to be part of the repo, rather than in `$XDG_CONFIG_HOME` or wherever.

We could support both options, and do any of the following:

* Merge the files, giving more priority to the project settings, or;
* Use the project settings if exists, or fallback the system settings (if exists)

But there is more value in adding support on project settings first.

---

I have a question: do all Dhall implementation have the same flags? (I think not). Having this config file makes me think that every implementation should support those flags. Because of that, should we standardize flags as well?

**Comment by Nadrieril** [comment]
Do other dhall implementations even have binaries that could take those flags? `dhall-rust` is only meant to be used to make an application that reads dhall files so flags wouldn't make sense.
However, standardizing `.dhall-config.dhall` would make a lot of sense.

**Comment by Gabriel439** [comment]
Oh yeah, this issue should probably be specific to the Haskell implementation.  I didn't realize that this was in the `dhall-lang` repository.  However, we could support the same functionality across other implementations, too, if they were interested.

**Comment by mheiber** [comment]
> Oh yeah, this issue should probably be specific to the Haskell implementation. I didn't realize that this was in the `dhall-lang` repository. However, we could support the same functionality across other implementations, too, if they were interested.

I'd hope that eventually the behavior would be specified so that everyone will get the same behavior as their teammates, regardless of choice of toolchain. I understand if it makes sense to experiment in the Haskell implementation first, though. Should I close this issue and open one there?

**Comment by Gabriel439** [comment]
So formatting preference could be reasonably agnostic of the language binding, but the explain option is definitely Haskell-specific.

I also agree with @philandstuff that `XDG_CACHE_HOME` falling back to `~/.config` is probably where I'd recommend to save something like this.

Probably for now we should start with the assumption that it's specific to the Haskell implementation and name the configuration file `dhall-haskell.config`

**Comment by mheiber** [comment]
> I also agree with @philandstuff that `XDG_CACHE_HOME` falling back to `~/.config` is probably where I'd recommend to save something like this.
> 

fwiw, this wouldn't meet my use case of easing collaboration with team members. Ideally, it would be possible to:
- have different projects with different settings
- check the settings into a repo, so that other team members have a 0-config way of automatically agreeing on stuff

**Comment by Gabriel439** [comment]
@mheiber: So I suppose that a local config is okay, but I think the tool should also fall back to looking in `XDG_CACHE_HOME` / `~/.config` as well if no local config file is present

**Comment by mheiber** [comment]
> @mheiber: So I suppose that a local config is okay, but I think the tool should also fall back to looking in `XDG_CACHE_HOME` / `~/.config` as well if no local config file is present

How about the following algorithm?
- recursively search up parent directories for dhall-haskell.dhall, stopping at `/`
- if not found, look in `$XDG_CACHE_HOME`
- if not found, look in ~/.config`

**Comment by Gabriel439** [comment]
@mheiber: Yeah, that should be fine

**Comment by philandstuff** [comment]
It should be `XDG_CONFIG_HOME`, not `XDG_CACHE_HOME`. 

But I’m actually having second thoughts. The thread is taking about loading the first config file it finds, but that would be highly surprising for me. I’d expect overrides to happen on a per-config-item basis, rather than a whole file. If I enable `—explain` in my home directory config file, I don’t want it disabled again just because I’m editing a project which prefers ASCII formatting. 

At this stage, I’m actually leaning more towards keeping things simple: let’s just standardize project-local config files for now, and if desired let’s have a future PR to allow `XDG_CONFIG_HOME`. This keeps changes to the standard small and focused and keeps our options open for the future. (I’m actually in favour of user config files but I have quite a lot of opinions here and I don’t want them to bog down the core of this PR if they don’t have to.)

**Comment by Gabriel439** [comment]
Yeah, that would be fine, too

**Comment by Nadrieril** [comment]
Is the config file to be read from the working directory or from the directory of the file we are loading? So far, dhall had the nice property that if I point it to a file, imports get resolved relative to that and the CWD does not matter.
I don't know which I would prefer; just pointing it out so that we're aware there's a design choice there. If we add more options to the config file later (like say setting import headers), this difference might matter.

**Comment by Gabriel439** [comment]
@Nadrieril: I think it should be relative to the configuration file's path

**Comment by joneshf** [comment]
Another way to address this issue would be to only have ascii syntax. This isn't the first time someone has mentioned unicode syntax being a concern (either formally in a GitHub issue, or informally between people). It doesn't seem like it will be the last time either.

**Comment by Nadrieril** [comment]
> @Nadrieril: I think it should be relative to the configuration file's path

I am confused by this answer. My question is, when reading `/path/to/foo/file.dhall`, and the current directory is `/path/to/bar`, should we look for the config file in `/path/to/foo/.dhall-config.dhall` or `/path/to/bar/.dhall-config.dhall`? I _think_ you mean `/path/to/foo/.dhall-config.dhall` but I'm not sure.

**Comment by Gabriel439** [comment]
@Nadrieril: Oh, if you mean what the search path would be for locating the configuration file, then I'm not sure which one I prefer.

**Comment by mheiber** [comment]
If we decide to have the search path be relative to the file being formatted, then how would `--transitive` work?
- relative to the start file. A downside of this approach is that in a monorepo with shared transitive deps people might take turns reformatting each other's config files.
- for each file discovered in the walk, relative to that file

**Comment by Gabriel439** [comment]
@mheiber: Probably relative to the start file

**Comment by basile-henry** [comment]
> I’d expect overrides to happen on a per-config-item basis, rather than a whole file. If I enable —explain in my home directory config file, I don’t want it disabled again just because I’m editing a project which prefers ASCII formatting.

A slight more complicated solution would be to make the config file be a function:

```dhall
-- .dhall-config.dhall
\(config : DhallConfig) ->

config with
  formatting.ascii = True
```

The definition of `DhallConfig` would be whatever the dhall binary used supports.

The various places where such config could be are then loaded in reverse order of importance (least important to most important), so that the most important override has precedence:

- From `~/.config`
- From `$XDG_CONFIG_HOME`
- From `/` all the way down to `$PWD`

**Comment by mheiber** [comment]
I have a working version here if anyone wants to try it out: https://github.com/dhall-lang/dhall-haskell/compare/master...mheiber:config-file?expand=1

I implemented this search strategy for dhall-config.dhall:
- cwd if reading from stdout
- current file if targeting on a single file
- first file if using --transitive

**Comment by mheiber** [comment]
I have a draft PR with this feature if anyone wants to try it out: https://github.com/dhall-lang/dhall-haskell/pull/2085

It implements this search strategy for dhall-config.dhall:
- cwd if reading from stdout
- directory of the targeted file if targeting a single file
    - when --transitive is passed, we still only pay attention to the directory of the first file

**Comment by PierreR** [comment]
@mheiber out of curiosity isn't this problem solved by defining an alias per folder/project with something like [direnv](https://direnv.net/) ?

Also, one place where a config file could be quite useful is `dhall-openapi` (https://github.com/dhall-lang/dhall-haskell/issues/2046)

**Comment by mheiber** [comment]
> @mheiber out of curiosity isn't this problem solved by defining an alias per folder/project with something like [direnv](https://direnv.net/) ?
> 
> Also, one place where a config file could be quite useful is `dhall-openapi` ([dhall-lang/dhall-haskell#2046](https://github.com/dhall-lang/dhall-haskell/issues/2046))

@PierreR thanks for this suggestion. I'm not sure if it would meet my use case but perhaps it will be good enough for most people.

 I'm looking for a low-barrier way to adopt Dhall. Ideally I'd put some dhall files in the repo and stuff would "just work"  the same for all team members who use the repo. Several team member didn't like the unicode lambdas and I was expecting to be able to do something to tell `dhall format` to stop unicodifying in a way that our text editors would all respect. It doesn't matter if I make a Makefile with its own format command if the Dhall VSCode extension still formats with default settings.

If the solution is to use `direnv` then, as someone trying to bring dhall to a team, I now have to sell either:
- programmable configuration *and* programming with non-ascii characters
- OR programmable configuration *and* `direnv`

So it doesn't sound like `direnv` would solve the problem I had, but maybe it would solve other people's problems.
I scanned the README and didn't actually see how this would work. Would I be aliasing `dhall format` to `dhall --format ascii`? Would the VSCode plugin's `Dhall format` action pick this up?

**Comment by Nadrieril** [comment]
I don't know if direnv does aliases, but we can do that with env vars. Since this seems specific to `dhall format` we don't need a cross-impl solution, and something simple like `DHALL_DEFAULT_OPTIONS="--ascii"` could do the job.

**Comment by mheiber** [comment]
> I don't know if direnv does aliases, but we can do that with env vars. Since this seems specific to `dhall format` we don't need a cross-impl solution, and something simple like `DHALL_DEFAULT_OPTIONS="--ascii"` could do the job.

Will this work with VSCode's format command? I don't think so, unless VSCode is started from a shell with the env var set. 

Similar issues arise for terminal-based editors

**Comment by PierreR** [comment]
@Nadrieril I like your suggestion. I have just checked `direnv` support for aliases and indeed it does not seem [straightforward](https://github.com/direnv/direnv/issues/109).

@mheiber there is a direnv extension for vscode, so I believe @Nadrieril suggestion would work nicely. Also `direnv` is a very generic tool and there is almost no overhead to introduce it. This is unlike `Dhall` which is much harder to "sell".

**Comment by mheiber** [comment]
@PierreR the issue you linked explains why `direnv`does *not* support aliases. It was marked as a duplicate of [the open `direnv` issue to support aliases and functions](https://github.com/direnv/direnv/issues/73).

**Comment by mheiber** [comment]
I updated the PR description to add Prior Art

**Comment by mheiber** [comment]
I updated the PR description to add Prior Art: every configurable code formatter I could find has a config file format.

I understand that there may be some way to cobble things together with other tools, rather than have dhall find its own config file. But what is the *advantage* of doing that?

**Comment by PierreR** [comment]
I agree that the `alias` option is a bit of a workaround but I feel that environment variables would be well suited for this use case.
It is much simpler to implement and you don't need to agree on where your configuration file(s) should be stored.

It is quite common for tools to receive configuration from 'command line', 'file config' and 'environment variables' together.

FWIW `vscode` should support setting env variable per workspace soonish: https://github.com/microsoft/vscode/issues/69233. That would then cleanly solve your use case without the need of an extra tool such as direnv.

That said, if a config file is added, I would not mind at all. I just hope all dhall executables (such as `dhall-openapi`, ...) would share the same file(s). As I said, that could be useful particularly for more complex configurations (for instance a long list of filtered paths).

**Comment by mheiber** [comment]
@PierreR 

> That said, if a config file is added, I would not mind at all. I just hope all dhall executables (such as `dhall-openapi`, ...) would share the same file(s).

That's what I had in mind, too. But I learned since opening this issue that 'dhall format' isn't part of dhall-lang, strictly-speaking, it seems like its own tool that happens to be in the same codebase as `dhall-haskell`, which implements the dhall language.

It seems to me that the use case for formatting defaults is distinct from other stuff one might want in a config file. `dhall-openapi` probably won't need to know about `--ascii`.

Would `dhall-format.dhall` be a better name for the config file than `dhall-config.dhall`? This would make the purpose clearer.

The PR currently implements `explain`/`concise` as well, which seems a bit different to me. There's no reason to standardize a team on error message verbosity, but it does seem important to standardize formatting. If we use `dhall-format.dhall` for formatting stuff, that leaves it open to do what's best for the core dhall functionality, be it a future `dhall-config.dhall` or some other file.

**Comment by Gabriel439** [comment]
@mheiber: You don't want to limit it to `dhall format`.  Other `dhall` subcommands also support `--ascii` and `--explain`  It should probably be called `dhall-haskell.dhall`

**Comment by mheiber** [comment]
There are a lot of interesting ideas in this discussion, thanks for everyone's work here.

It doesn't look like anyone else has hit the same problem I have, so it may be hard to find a solution that works well for my use case, as well as the unknown use cases as Dhall evolves.

One thing we could do is close the issue for now and then gauge interest in this year's [Dhall Survey](http://www.haskellforall.com/2020/02/dhall-survey-results-2019-2020.html) - it's only a few months away–would that work?

**Comment by Gabriel439** [comment]
@mheiber: Part of the reason that I'm partial to the `direnv`-based solution is that `direnv` seems to be a pretty mainstream tool that wouldn't be difficult to introduce to a team (certainly easier than introducing Dhall).  Also, the `direnv`-based solution with a few small tweaks would also work for other Dhall implementations (e.g. `dhall-rust`), rather than having to implement a separate configuration file format for each implementation.

**Comment by mheiber** [comment]
> @mheiber: Part of the reason that I'm partial to the `direnv`-based solution is that `direnv` seems to be a pretty mainstream tool that wouldn't be difficult to introduce to a team (certainly easier than introducing Dhall). Also, the `direnv`-based solution with a few small tweaks would also work for other Dhall implementations (e.g. `dhall-rust`), rather than having to implement a separate configuration file format for each implementation.

`direnv` [does not support aliases](https://github.com/direnv/direnv/issues/73). So is the `direnv`-based solution to wrap `dhall` in another executable and overwrite $PATH to point to this alternative executable?

**Comment by Gabriel439** [comment]
@mheiber: What if we made the behavior configurable via environment variables?

**Comment by mheiber** [comment]
> @mheiber: What if we made the behavior configurable via environment variables?

That would probably work. I don't think it's a great experience, though. Everyone's editors and shells would have to be direnv-aware in order to get the homogeneity I'm looking for. 

I don't think I'd use such a feature. But if there's a lot of demand for it, I'm happy to update the PR to use environment variables instead.

**Comment by mheiber** [comment]
If we're going with environment variables, do we want one for `explain` and one for `ascii`? Or just `ascii`?

Will it be something like case-insensitive `DHALL_FORMAT=ascii   / DHALL_FORMAT=plain` or `DHALL_HASKELL_ASCII=true` / false?

**Comment by mheiber** [comment]
If we're going with environment variables, would we start with just a formatting env var and see if we want an `explain` one later?

If so, how is `DHALL_FORMAT_ASCII=<boolean>`? We would keep the behavior where the env var can be overridden with the command-line flag `--plain`.

**Comment by Gabriel439** [comment]
@mheiber: Perhaps name the environment variable `DHALL_CHARSET` (or something similar), which could be set to `ASCII` or `UNICODE`, both to avoid boolean blindness and also to support other potential options in the future (e.g. a formatting mode where the interpreter tries to preserve the file's character set)

**Comment by Gabriel439** [comment]
@mheiber: Actually, talking about this more, why not make `dhall format` try to preserve the file's character set by default, instead of changing to an opinionated character set?  And if a file has an ambiguous character set (i.e. `→` and `->` within the same file) only then does it prefer one or the other?

**Comment by basile-henry** [comment]
> preserve the file's character set by default

That sounds like a great default! If there's mixed style in the file I think defaulting to using Unicode formatting is fine, a user is less likely to add a Unicode character by mistake and for user that do prefer Unicode, they might still prefer typing in ASCII (and have dhall format convert it).

**Comment by Nadrieril** [comment]
If we want this to be useful, I imagine that in the mixed case we want to transform everything to unicode. That's because as a user if there's a file full of dhall unicode symbols, I want to be able to write a `->` and have `dhall format` convert it.
However that means that if I prefer ascii and someday I add a comment in e.g. French that uses non-ascii characters, suddenly all the file gets converted to unicode symbols. So I think we want to ignore comments

**Comment by SiriusStarr** [comment]
> Actually, talking about this more, why not make dhall format try to preserve the file's character set by default, instead of changing to an opinionated character set? 

One problem I foresee with this:  How does one **start** a new unicode Dhall file?  You'd then have to go copy-paste (or input by some more arcane system) a unicode character in rather than just writing and letting `dhall format` handle it as one currently does.

**Comment by basile-henry** [comment]
> How does one start a new unicode Dhall file?

We already have a flag `--ascii` to specify the character set. We could add `--unicode`. `dhall format` would use the flag if specified, otherwise try to infer the character set from the file.

**Comment by Gabriel439** [comment]
@Nadrieril: I'd like to clarify that I meant for the character set detection to only consider language punctuation (e.g. `→`, `∀`) and not consider the character set of identifiers, strings, or comments

**Comment by SiriusStarr** [comment]
> We already have a flag `--ascii` to specify the character set. We could add `--unicode`. `dhall format` would use the flag if specified, otherwise try to infer the character set from the file.

Yeah, that makes sense, as long as that flag can be set in the language server plugin (not that it would be the end of the world to have to run format in the terminal once).

**Comment by basile-henry** [comment]
I have a possible implementation for detecting character set over at https://github.com/dhall-lang/dhall-haskell/pull/2108 if anyone wants to try it and see if there's anything wrong with it :sweat_smile:

**Comment by mheiber** [comment]
> I have a possible implementation for detecting character set over at [dhall-lang/dhall-haskell#2108](https://github.com/dhall-lang/dhall-haskell/pull/2108) if anyone wants to try it and see if there's anything wrong with it 😅

Interesting idea, but I personally prefer determinism and explicitness.

**Comment by mheiber** [comment]
fwiw, I prefer explicitness and predictability–simple answers to simple questions like how a file will be formatted.

Re comment https://github.com/dhall-lang/dhall-lang/issues/1086#issuecomment-738126735: this does seem to be a serious concern, since the most common workflow is to input text as ascii and then autoformat it into unicode ([from last year's survey](http://www.haskellforall.com/2020/02/dhall-survey-results-2019-2020.html)).

**Comment by basile-henry** [comment]
This solution does leave the possibility to be explicit since you have both flags available in the CLI. But having a config file or using an environment variable as well is definitely still possible.

I think it is predictable, or at least deterministic. But you're right, it might not follow the principle of least surprise. It might also make the Unicode syntax less discoverable to new users.

I believe there is a precedent for formatters to use the current format of a file to inform how to format it (for example [`elm-format`](https://github.com/avh4/elm-format), or [`ormolu`](https://github.com/tweag/ormolu)). This is often used to decide when to format something over a single line or over multi-lines. So I think it makes sense to extend this approach to which character set is used.

> the most common workflow is to input text as ascii and then autoformat it into unicode

I think that is a great workflow, but I do wonder if it is the most common simply because the current default for `dhall format` is Unicode. With this character set detection you do have to take an extra step to get Unicode formatting, either by formatting it once with the CLI flag `--unicode` or by inserting a Unicode bit of syntax manually (or copy/paste).

I will agree with you that neither trick is ideal, which is why I think there is still room for a solution based on a config (via a file or an environment variable).

I do have another idea, though definitely a bit more of a hack (:sweat_smile:) to get `dhall format` to use a specific character set specified by the user from the file itself. We could use a _magic_ comment in the input, for example `-- dhall format unicode` that would tell `dhall format` to use Unicode. After the first format introducing Unicode syntax the comment could be removed. This is similar to what other tools do (for example [`shellcheck`](https://github.com/koalaman/shellcheck) or [`hlint`](https://github.com/ndmitchell/hlint)) to disable a warning, an error, or a lint.

**Comment by klardotsh** [comment]
I'd love to adopt Dhall (again) at my current org, but one of the pain points I ran into while piloting it at $LASTJOB was around what I didn't at the time realize was this ticket. (There, I just kinda hand-fixed the problems as they came up; that's a mistake I'd rather not walk myself back into again)

My company's eng teams collectively represent probably 5-6 different editors, a few shells, Docker and bare metal usecases, MacOS and Linux both, and probably a myriad of other quirks and variables I'm not considering. Amongst these permutations (most of which work for the developers and we're not, on the SRE/platform eng side of things, interested in mandating _too_ much about a workflow unless necessary) support all (or sometimes any) of the currently-proposed solutions of:

- env vars at all (at least in a reproducible-across-a-team way that doesn't involve tons of manual setup - this is mostly the GUI IDE folks, not really a problem for the Vim folks)
- env vars in a managed fashion a-la `direnv` (`direnv` is already on our wishlist to roll out for CLI tools so it's probably the most palatable of this set, but is limited to the CLI tools executed directly, and won't help those looking to use in-editor auto-formatting)
- LSP at all
- plumbing LSP-local config variables around in the editor

I would love to help reboot this thread towards a `.dhall-haskell.dhall` config file I can just drop in the org's Git repos and never think about again. What can I do to help this happen? Is there interest in accepting a PR (possibly just a minimum viable PR that isn't boil-the-ocean complete, just due to my time availability) for such? Alternatively, is there a workaround that's recommended to (in some other way) be able to set flags for all users within an org (and in the CI system) that doesn't rely on the existence of LSP-local config var plumbing?

Cheers!

**Comment by Gabriella439** [comment]
If you only propose changing the behavior of the Haskell implementation then it's very easy to do since it only requires my approval and I'm very permissive about accepting Haskell-specific UX enhancements

I think my main question is what you'd like to specify in this config file since the formatter should preserve formatting by default now

**Comment by klardotsh** [comment]
Ideally I'd like to set `--ascii` (to ensure that even if someone copy-pastes a unicode example from StackOverflow or something, it's auto-corrected to use ASCII, and CI will be able to lint against that known spec), but there was also an interesting proposal in https://github.com/dhall-lang/dhall-lang/issues/1086#issuecomment-713717854 about being able to set JSON output parameters in such a file, which I might explore a bit if time allows.

**Comment by Gabriella439** [comment]
Yeah, in that case I'd accept a pull request to add `dhall-haskell`-specific configuration file if the intention is to control behavior specific to `dhall-haskell` (such as formatting)
